### PR TITLE
Getting the module rulling

### PR DESCRIPTION
**Pull request by cableman**
#### Link to ticket

https://leantime.itkdev.dk/#/tickets/showTicket/1452

#### Description

This is Drupal module to communicate with LLM's.

#### Screenshot of the result

Screenshot from 2024-05-28 14-59-38

#### Checklist

- [x] My code passes our static analysis suite.
- [x] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change
should be exempt from the list.

#### Additional comments or questions

If you have any further comments or questions for the reviewer please add them here.

### CODE REVIEW

File: llm_services.info.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+name: "Large language model services"
+description: 'Large language module services to communicat with the models.'
```
Comment (jekuaitk) [inline llm_services.info.yml]:
```suggestion
description: 'Large language module services to communicate with the models.'
```

File: llm_services.links.task.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+llm_services.plugin_settings_tasks:
+  title: 'LLM Services settings'
```
Comment (jekuaitk) [inline llm_services.links.task.yml]:
```suggestion
  title: 'LLM services settings'
```

File: llm_services.routing.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+llm_services.plugin_settings_local_tasks:
+  path: '/admin/config/llm_services/settings/{type}'
+  defaults:
+    _controller: '\Drupal\llm_services\Controller\LocalTasksController::dynamicTasks'
+    _title: 'LLM Services settings'
```
Comment (jekuaitk) [inline llm_services.routing.yml]:
```suggestion
    _title: 'LLM services settings'
```

File: src/Client/Ollama.php [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  private string $parserCache = '';
+
+  /**
+   * Default constructor.
+   *
+   * @param string $url
+   *   The URL of the Ollama server.
+   * @param int $port
+   *   The port that Ollama is listing on.
```
Comment (jekuaitk) [inline src/Client/Ollama.php]:
```suggestion
   *   The port that Ollama is listening on.
```

File: src/Drush/Commands/LlmServicesCommands.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+   * Try out completion with a model.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   * @throws \Drupal\llm_services\Exceptions\CommunicationException
+   */
+  #[CLI\Command(name: 'llm:model:completion', aliases: ['llm-completion'])]
+  #[CLI\Argument(name: 'provider', description: 'Name of the provider (plugin).')]
+  #[CLI\Argument(name: 'name', description: 'Name of the model to use.')]
+  #[CLI\Argument(name: 'prompt', description: 'The prompt to generate a response for.')]
+  #[CLI\Usage(name: 'llm:model:completion ollama llama2 "Why is the sky blue?"', description: 'Install LLama2 modul in Ollama')]
```
Comment (jekuaitk) [inline src/Drush/Commands/LlmServicesCommands.php]:
```suggestion
  #[CLI\Usage(name: 'llm:model:completion ollama llama2 "Why is the sky blue?"', description: 'Prompt LLama2')]
```

File: src/Exceptions/LLMException.php [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace Drupal\llm_services\Exceptions;
+
+/**
+ * Base execution that all other exceptions should extend.
```
Comment (jekuaitk) [inline src/Exceptions/LLMException.php]:
```suggestion
 * Base exception that all other exceptions should extend.
```

File: src/Form/SettingsForm.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    $options = array_map(function ($plugin) {
+      /** @var \Drupal\Core\StringTranslation\TranslatableMarkup $title */
+      $title = $plugin['title'];
+      return $title->render();
+    }, $plugins);
+
+    $form['provider'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Log provider'),
+      '#description' => $this->t('Select the logger provider you which to use'),
```
Comment (jekuaitk) [inline src/Form/SettingsForm.php]: What does this have to do with 'log'? Should it just be

```suggestion
      '#title' => $this->t('Provider'),
      '#description' => $this->t('Select the provider you wish to use'),
```

File: src/Model/MessageRoles.php [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+ * is in the conversation. E.g., "You are a friendly assistant here to help."
+ *
+ * @enum MessageRoles
+ *
+ * @type string
+ */
+enum MessageRoles: string {
+  case Assistant = 'assistant';
+  case System = "system";
+  case User = 'user';
```
Comment (jekuaitk) [inline src/Model/MessageRoles.php]:
```suggestion
  case Assistant = 'assistant';
  case System = 'system';
  case User = 'user';
```

File: src/Plugin/Derivative/LocalTask.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+   * @throws \ReflectionException
+   */
+  public function getDerivativeDefinitions($base_plugin_definition): array {
+    $plugins = $this->providerManager->getDefinitions();
+    ksort($plugins);
+
+    // Sadly, it seems that it is not possible to just invalidate the
+    // deriver/menu cache stuff. To get the local tasks menu links. So instead
+    // of clearing all caches on settings save to only show selected plugins, we
+    // show em all.
```
Comment (jekuaitk) [inline src/Plugin/Derivative/LocalTask.php:41]: 👍

File: src/Drush/Commands/LlmServicesCommands.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+use Drupal\llm_services\Model\Payload;
+use Drupal\llm_services\Plugin\LLModelProviderManager;
+use Drush\Attributes as CLI;
+use Drush\Commands\DrushCommands;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Drush commands to talk with LLM provider (mostly for testing).
+ */
+final class LlmServicesCommands extends DrushCommands {
```
Comment (jekuaitk) [inline src/Drush/Commands/LlmServicesCommands.php]: See general comment.

**Review by jekuaitk — CHANGES_REQUESTED** [review]
Looks good, but we should be consistent with the use of abbreviations in class names etc, i.e. `LLM` vs `Llm` vs `LLModel` vs `LLModels`.

**Comment by cableman** [comment]
I have tried to clean up the namespaces to better match what I try to do.

File: src/Commands/ModelChatCommand.php [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+      ->addOption(
+        name: 'temperature',
+        mode: InputOption::VALUE_REQUIRED,
+        description: 'The temperature of the model. Increasing the temperature will make the model answer more creatively.',
+        default: '0.8'
+      )
+      ->addOption(
+        name: 'top-k',
+        mode: InputOption::VALUE_REQUIRED,
+        description: 'Reduces the probability of generating nonsense. A higher value (e.g. 100) will give more diverse answers.',
```
Comment (jekuaitk) [inline src/Commands/ModelChatCommand.php:68]: I read the first sentence as if a higher value reduce nonsense, but the second sentence reveals the opposite.

File: src/Commands/ModelChatCommand.php [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    $msg = new Message();
+    $msg->role = MessageRoles::System;
+    $msg->content = $systemPrompt;
+    $payLoad->messages[] = $msg;
+
+    /** @var \Symfony\Component\Console\Helper\HelperInterface $helper */
+    $helper = $this->getHelper('question');
+    $question = new Question('Message: ', '');
+
+    // Keep cheating with the user. Not optimal, but okay for now.
```
Comment (jekuaitk) [inline src/Commands/ModelChatCommand.php]:
```suggestion
    // Keep chatting with the user. Not optimal, but okay for now.
```

File: src/Commands/ModelChatCommand.php [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+   * {@inheritDoc}
+   */
+  protected function execute(InputInterface $input, OutputInterface $output): int {
+    $providerName = $input->getArgument('provider');
+    $name = $input->getArgument('name');
+
+    $systemPrompt = $input->getOption('system-prompt');
+    $temperature = $input->getOption('temperature');
+    $topK = $input->getOption('top-k');
+    $topP = $input->getOption('top-p');
```
Comment (jekuaitk) [inline src/Commands/ModelChatCommand.php:89]: Should we add some checks for allowed values of these?

I'm not fully aware of what these values are/should be, but it seems that `$temperature` and `$topP` should be a decimal between 0-1 and `$topK` a positive integer?

File: src/Commands/ModelCompletionCommand.php [hunk 14]
```diff
@@ -0,0 +1,109 @@
+<?php
```
Comment (jekuaitk) [inline src/Commands/ModelCompletionCommand.php:1]: Same comments as in `ModelChatCommand`

File: src/Commands/ProviderListCommand.php [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    parent::__construct();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  protected function configure(): void {
+    $this
+      ->setName('llm:provider:list')
+      ->setDescription('Install model in provider')
```
Comment (jekuaitk) [inline src/Commands/ProviderListCommand.php]:
```suggestion
      ->setDescription('Lists installed models in provider')
```

**Review by jekuaitk — CHANGES_REQUESTED** [review]

File: src/Commands/ModelChatCommand.php (hunk 11 above)
Comment (cableman) [reply src/Commands/ModelChatCommand.php:68]: yes, it is taken directly from Ollama documentation...

File: src/Commands/ModelChatCommand.php (hunk 13 above)
Comment (cableman) [reply src/Commands/ModelChatCommand.php:89]: This is the recommended values... but not hard limits... best I can do is checking they are numbers

Comment (cableman) [reply src/Commands/ModelChatCommand.php:89]: Added simple validation

**Review by jekuaitk — APPROVED** [review]
